File: states.js
/**
 * Rotas para Estados
 * 
 * Gerencia todas as operações relacionadas aos estados fictícios
 */

const express = require('express');
const router = express.Router();
const State = require('../models/State');
const database = require('../config/database');

/**
 * POST /api/states
 * Cria um novo estado
 */
router.post('/', async (req, res) => {
  try {
    const { name, region, government_type, player_id } = req.body;

    // Validar dados obrigatórios
    if (!name || !region || !government_type) {
      return res.status(400).json({
        error: true,
        message: 'Nome, região e tipo de governo são obrigatórios',
        required: ['name', 'region', 'government_type']
      });
    }

    // Verificar se já existe um estado com esse nome
    const existingStates = await database.getStates();
    const nameExists = existingStates.some(state => 
      state.name.toLowerCase() === name.toLowerCase()
    );

    if (nameExists) {
      return res.status(409).json({
        error: true,
        message: 'Já existe um estado com esse nome. Escolha outro nome!',
        suggestion: 'Que tal adicionar um número ou modificar um pouco o nome?'
      });
    }

    // Criar novo estado
    const stateData = {
      name: name.trim(),
      region,
      government_type,
      player_id: player_id || null,
      indicators: {
        economia: 50,
        educacao: 50,
        saude: 50,
        seguranca: 50,
        cultura: 50,
        satisfacao_popular: 50,
        corrupcao: 50
      }
    };

    const newState = new State(stateData);
    
    // Validar estado
    const validation = newState.validate();
    if (!validation.isValid) {
      return res.status(400).json({
        error: true,
        message: 'Dados inválidos para criação do estado',
        errors: validation.errors
      });
    }

    // Salvar no banco
    const savedState = await database.createState(newState.toJSON());

    res.status(201).json({
      success: true,
      message: `🎉 Estado ${savedState.name} criado com sucesso!`,
      state: savedState,
      tips: [
        'Todos os indicadores começam em 50/100',
        'Tome decisões políticas para melhorar seus indicadores',
        'Acompanhe os rankings para ver como você está se saindo!'
      ]
    });

  } catch (error) {
    console.error('Erro ao criar estado:', error);
    res.status(500).json({
      error: true,
      message: 'Erro interno do servidor ao criar estado',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

/**
 * GET /api/states/:id
 * Busca um estado específico por ID
 */
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const state = await database.getStateById(id);
    
    if (!state) {
      return res.status(404).json({
        error: true,
        message: 'Estado não encontrado',
        suggestion: 'Verifique se o ID está correto ou se o estado ainda existe'
      });
    }

    // Criar objeto State para calcular informações adicionais
    const stateObj = new State(state);

    res.json({
      success: true,
      state: stateObj.toJSON(),
      lastUpdate: state.updated_at
    });

  } catch (error) {
    console.error('Erro ao buscar estado:', error);
    res.status(500).json({
      error: true,
      message: 'Erro interno do servidor ao buscar estado',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

/**
 * GET /api/states
 * Lista todos os estados (com paginação opcional)
 */
router.get('/', async (req, res) => {
  try {
    const { page = 1, limit = 10, region, government_type, sort_by = 'created_at' } = req.query;
    
    let states = await database.getStates();

    // Filtros
    if (region) {
      states = states.filter(state => state.region === region);
    }

    if (government_type) {
      states = states.filter(state => state.government_type === government_type);
    }

    // Ordenação
    states.sort((a, b) => {
      switch (sort_by) {
        case 'name':
          return a.name.localeCompare(b.name);
        case 'overall_score':
          const scoreA = new State(a).calculateOverallScore();
          const scoreB = new State(b).calculateOverallScore();
          return scoreB - scoreA;
        case 'created_at':
        default:
          return new Date(b.created_at) - new Date(a.created_at);
      }
    });

    // Paginação
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + parseInt(limit);
    const paginatedStates = states.slice(startIndex, endIndex);

    // Adicionar informações calculadas
    const statesWithInfo = paginatedStates.map(state => {
      const stateObj = new State(state);
      return stateObj.toJSON();
    });

    res.json({
      success: true,
      states: statesWithInfo,
      pagination: {
        current_page: parseInt(page),
        per_page: parseInt(limit),
        total: states.length,
        total_pages: Math.ceil(states.length / limit),
        has_next: endIndex < states.length,
        has_prev: page > 1
      },
      filters: {
        region,
        government_type,
        sort_by
      }
    });

  } catch (error) {
    console.error('Erro ao listar estados:', error);
    res.status(500).json({
      error: true,
      message: 'Erro interno do servidor ao listar estados',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

/**
 * PATCH /api/states/:id/decision
 * Aplica uma decisão política ao estado
 */
router.patch('/:id/decision', async (req, res) => {
  try {
    const { id } = req.params;
    const { decision_id, option_id } = req.body;

    if (!decision_id || !option_id) {
      return res.status(400).json({
        error: true,
        message: 'ID da decisão e da opção são obrigatórios',
        required: ['decision_id', 'option_id']
      });
    }

    // Buscar estado
    const state = await database.getStateById(id);
    if (!state) {
      return res.status(404).json({
        error: true,
        message: 'Estado não encontrado'
      });
    }

    // Buscar decisão
    const decisions = await database.getDecisions();
    const decision = decisions.find(d => d.id === decision_id);
    
    if (!decision) {
      return res.status(404).json({
        error: true,
        message: 'Decisão não encontrada'
      });
    }

    // Buscar opção
    const option = decision.options.find(o => o.id === option_id);
    if (!option) {
      return res.status(404).json({
        error: true,
        message: 'Opção de decisão não encontrada'
      });
    }

    // Verificar cooldown (24 horas entre decisões)
    if (state.last_decision_at) {
      const lastDecision = new Date(state.last_decision_at);
      const now = new Date();
      const hoursSinceLastDecision = (now - lastDecision) / (1000 * 60 * 60);
      
      const cooldownHours = parseInt(process.env.DECISION_COOLDOWN_HOURS) || 24;
      
      if (hoursSinceLastDecision < cooldownHours) {
        const remainingHours = Math.ceil(cooldownHours - hoursSinceLastDecision);
        return res.status(429).json({
          error: true,
          message: `Calma aí, chefe! Você precisa esperar ${remainingHours}h para tomar outra decisão.`,
          cooldown: {
            remaining_hours: remainingHours,
            next_decision_at: new Date(lastDecision.getTime() + (cooldownHours * 60 * 60 * 1000)).toISOString()
          }
        });
      }
    }

    // Aplicar efeitos da decisão
    const stateObj = new State(state);
    stateObj.applyDecisionEffects(option.effects);
    stateObj.addDecisionToHistory(decision_id, option_id, option.effects);

    // Salvar estado atualizado
    const updatedState = await database.updateState(id, stateObj.toJSON());

    // Gerar texto de resultado
    const Decision = require('../models/Decision');
    const resultText = Decision.generateResultText(option, option.effects);

    res.json({
      success: true,
      message: 'Decisão aplicada com sucesso!',
      state: updatedState,
      decision: {
        id: decision_id,
        title: decision.title,
        chosen_option: option.text,
        effects: option.effects
      },
      result: {
        text: resultText,
        new_indicators: updatedState.indicators,
        changes: option.effects
      }
    });

  } catch (error) {
    console.error('Erro ao aplicar decisão:', error);
    res.status(500).json({
      error: true,
      message: 'Erro interno do servidor ao aplicar decisão',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

/**
 * DELETE /api/states/:id
 * Remove um estado (apenas para desenvolvimento)
 */
router.delete('/:id', async (req, res) => {
  try {
    if (process.env.NODE_ENV !== 'development') {
      return res.status(403).json({
        error: true,
        message: 'Operação não permitida em produção'
      });
    }

    const { id } = req.params;
    const states = await database.getStates();
    const stateIndex = states.findIndex(state => state.id === id);

    if (stateIndex === -1) {
      return res.status(404).json({
        error: true,
        message: 'Estado não encontrado'
      });
    }

    const deletedState = states[stateIndex];
    states.splice(stateIndex, 1);
    await database.saveStates(states);

    res.json({
      success: true,
      message: `Estado ${deletedState.name} removido com sucesso`,
      deleted_state: deletedState
    });

  } catch (error) {
    console.error('Erro ao remover estado:', error);
    res.status(500).json({
      error: true,
      message: 'Erro interno do servidor ao remover estado',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

module.exports = router;

File: states.py
"""
Rotas para Estados - BrasilSim
"""
from flask import Blueprint, request, jsonify
from src.models.state import State
from src.models.decision import Decision
from src.models import db
import uuid

states_bp = Blueprint('states', __name__)

@states_bp.route('/states', methods=['POST'])
def create_state():
    """Criar novo estado"""
    try:
        data = request.get_json()
        
        # Validar dados obrigatórios
        if not data or not all(k in data for k in ["name", "region", "government"]):
            return jsonify({
                "error": "Dados obrigatórios: name, region, government"
            }), 400
        
        name = data["name"].strip()
        region = data["region"]
        government = data["government"]
        
        # Validar região
        valid_regions = ['Norte', 'Nordeste', 'Centro-Oeste', 'Sudeste', 'Sul']
        if region not in valid_regions:
            return jsonify({
                'error': f'Região deve ser uma de: {", ".join(valid_regions)}'
            }), 400
        
        # Validar governo
        valid_governments = ['Democracia', 'Tecnocracia', 'Coronelismo']
        if government not in valid_governments:
            return jsonify({
                'error': f'Governo deve ser um de: {", ".join(valid_governments)}'
            }), 400
        
        # Verificar se nome já existe
        existing_state = State.query.filter_by(name=name).first()
        if existing_state:
            return jsonify({
                'error': 'Já existe um estado com este nome'
            }), 409
        
        # Criar estado
        state = State.create_state(name, region, government)
        
        return jsonify({
            'success': True,
            'message': f'Estado {name} criado com sucesso!',
            'state': state.to_dict()
        }), 201
        
    except Exception as e:
        return jsonify({
            'error': f'Erro interno do servidor: {str(e)}'
        }), 500

@states_bp.route('/states/<state_id>', methods=['GET'])
def get_state(state_id):
    """Buscar estado por ID"""
    try:
        # Validar UUID
        try:
            uuid.UUID(state_id)
        except ValueError:
            return jsonify({'error': 'ID inválido'}), 400
        
        state = State.get_by_id(state_id)
        if not state:
            return jsonify({'error': 'Estado não encontrado'}), 404
        
        return jsonify({
            'state': state.to_dict()
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': f'Erro interno do servidor: {str(e)}'
        }), 500

@states_bp.route('/states', methods=['GET'])
def list_states():
    """Listar todos os estados"""
    try:
        states = State.get_all_states()
        return jsonify({
            'states': [state.to_dict() for state in states],
            'total': len(states)
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': f'Erro interno do servidor: {str(e)}'
        }), 500

@states_bp.route('/states/<state_id>/decision', methods=['PATCH'])
def apply_decision(state_id):
    """Aplicar decisão política ao estado"""
    try:
        # Validar UUID
        try:
            uuid.UUID(state_id)
        except ValueError:
            return jsonify({'error': 'ID inválido'}), 400
        
        data = request.get_json()
        if not data or 'decisionId' not in data or 'optionIndex' not in data:
            return jsonify({
                'error': 'Dados obrigatórios: decisionId, optionIndex'
            }), 400
        
        state = State.get_by_id(state_id)
        if not state:
            return jsonify({'error': 'Estado não encontrado'}), 404
        
        # Verificar cooldown
        if not state.can_make_decision():
            return jsonify({
                'error': 'Você deve aguardar 24 horas entre decisões'
            }), 429
        
        # Buscar decisão
        decision = Decision.query.get(data['decisionId'])
        if not decision:
            return jsonify({'error': 'Decisão não encontrada'}), 404
        
        # Validar opção
        options = decision.get_options()
        option_index = data['optionIndex']
        
        if option_index < 0 or option_index >= len(options):
            return jsonify({'error': 'Índice de opção inválido'}), 400
        
        selected_option = options[option_index]
        
        # Aplicar efeitos
        effects = selected_option.get('effects', {})
        state.update_indicators(effects)
        
        # Registrar decisão no histórico
        decision_record = {
            'decisionId': str(decision.id),
            'title': decision.title,
            'selectedOption': selected_option['text'],
            'effects': effects
        }
        state.add_decision(decision_record)
        
        # Salvar no banco
        db.session.commit()
        
        return jsonify({
            'message': 'Decisão aplicada com sucesso!',
            'state': state.to_dict(),
            'appliedEffects': effects
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({
            'error': f'Erro interno do servidor: {str(e)}'
        }), 500

@states_bp.route('/states/<state_id>/random-decision', methods=['GET'])
def get_random_decision(state_id):
    """Obter decisão aleatória para o estado"""
    try:
        # Validar UUID
        try:
            uuid.UUID(state_id)
        except ValueError:
            return jsonify({'error': 'ID inválido'}), 400
        
        state = State.get_by_id(state_id)
        if not state:
            return jsonify({'error': 'Estado não encontrado'}), 404
        
        # Verificar cooldown
        if not state.can_make_decision():
            return jsonify({
                'error': 'Você deve aguardar 24 horas entre decisões',
                'canMakeDecision': False
            }), 429
        
        # Buscar decisão aleatória
        decision = Decision.get_random_decision(state)
        if not decision:
            return jsonify({'error': 'Nenhuma decisão disponível'}), 404
        
        return jsonify({
            'decision': decision.to_dict(),
            'canMakeDecision': True
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': f'Erro interno do servidor: {str(e)}'
        }), 500

